### PR TITLE
feat(BindingBehavior): Support self binding behavior for event

### PR DESCRIPTION
**Pull request by bigopon**
aurelia/templating-binding#101

Tests are still failing (throttle, debounce, self) as aurelia-binding dependency is not updated

Usage (& test):
```html
<div click.delegate="doSomething($event.target.tagName.toLowerCase() === 'div') & self">
  <button click.delegate="wontdo">Clicking me won't trigger my div parent listener</button>
</div>
```

Where this can be useful:

Suggest we have a modal component with following markup:

```html
<div class="overlay">
  <div class="content-wrapper">
    <!-- ...... -->
  </div>
</div>
```

Without self binding behavior we need to add some extra stuff or separate dom path for `div.overlay` to perform action close on click the overlay.

With self binding behavior:

```html
<div class="overlay" click.delegate='close() & self'>
  <div class="content-wrapper">
    <!-- ...... -->
  </div>
</div>
```

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
Can you explain this feature further? @jdanyow Can you review?

**Comment by bigopon** [comment]
In this example

```html
<panel>
  <header mousedown.delegate='onMouseDown()'>
    <button class='settings'></button>
    <button class='close'></button>
  </header>
</panel>
```

Use case is I want when I mouse down on header, it start listening to `mousemove` to do drag. In the above markup, it requires to check if `event.target === header` before you do something else to make sure you don't start listening to `mousemove` when mouse down on the buttons.

* Rewritten with self binding behavior:

```html
<panel>
  <header mousedown.delegate='onMouseDown() & self'>
    <!-- mouse down on any of these buttons won't call onMouseDown -->
    <button class='settings'></button>
    <button class='close'></button>
  </header>
</panel>
```

The `self` binding behavior will help remove event stuff of of your business/ component logic so `onMouseDown` is ensured to only fire when user actually mouse down on header itself, not its descendants.

File: src/self-binding-behavior.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+function handleSelfEvent(event) {
+  let target = findOriginalEventTarget(event);
+  if (this.target !== target) return;
+  this.originalCallSource(event);
+}
+
+export class SelfBindingBehavior {
+  bind(binding, source) {
+    if (!binding.callSource || !binding.targetEvent) throw new Error('Self binding behavior only supports event.');
+    binding.originalCallSource = binding.callSource;
```
Comment (jdanyow) [inline src/self-binding-behavior.js]: instead of `originalCallSource` (which might conflict with someone else's binding behavior logic) try using `selfCallSource` or something else that is specific to this binding behavior.

File: src/self-binding-behavior.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    binding.callSource = handleSelfEvent;
+  }
+
+  unbind(binding, source) {
+    this.callSource = this.originalCallSource;
```
Comment (jdanyow) [inline src/self-binding-behavior.js]: shouldn't `this.` be `binding.` here?

File: src/self-binding-behavior.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    this.originalCallSource = null;
+  }
```
Comment (jdanyow) [inline src/self-binding-behavior.js]: `this.` -> `binding.`

File: test/debounce-binding-behavior.spec.js [hunk 4]
```diff
@@ -168,7 +169,7 @@ describe('DebounceBindingBehavior', () => {
       bindingEngine.observerLocator.eventManager,
       'mousemove',
       bindingEngine.parseExpression(`handleMouseMove($event) & debounce:${delay}`),
-      false,
+      delegationStrategy.none,
       false,
```
Comment (jdanyow) [inline test/debounce-binding-behavior.spec.js]: does this belong in another PR?

File: test/throttle-binding-behavior.spec.js [hunk 5]
```diff
@@ -143,7 +144,7 @@ describe('ThrottleBindingBehavior', () => {
       bindingEngine.observerLocator.eventManager,
       'mousemove',
       bindingEngine.parseExpression(`handleMouseMove($event) & throttle:${delay}`),
-      false,
+      delegationStrategy.none,
       false,
```
Comment (jdanyow) [inline test/throttle-binding-behavior.spec.js]: does this belong in another PR

Comment (bigopon) [inline test/throttle-binding-behavior.spec.js]: Hi, it was already merged but I don't know how to update my fork. Can you help please? I tried but couldn't find any clue why.

**Comment by EisenbergEffect** [comment]
@bigopon Thanks for the explanation. That sounds great.

File: test/throttle-binding-behavior.spec.js (hunk 5 above)
Comment (bigopon) [inline test/throttle-binding-behavior.spec.js]: Great, I read about rebase but not sure if it was the right thing to do, git documentation was quite a puzzle for me plus I'm still newbie. Tks.

Updates:
- Rebased
- Added test for all `delegationStrategy`
- Fixed unbind

**Comment by bigopon** [comment]
@jdanyow I've updated all deps. Tests are passing now

**Comment by jdanyow** [comment]
@bigopon this is looking really good. One last thing- could you add a section to this doc that explains how to use this feature?

https://github.com/aurelia/binding/blob/master/doc/article/en-US/binding-binding-behaviors.md

**Comment by bigopon** [comment]
@jdanyow sure, doing. But it seems to be more challenging than the writing code 😢
